### PR TITLE
Added DownloadItem .OriginalUrl

### PR DESCRIPTION
**Pull request by philipbeber**
This property is very valuable and I need it in the project I am working on right now!

Also:

Added a sidebar to the example app which shows the DownloadItems as received by CefSharp.Core/Internals/TypeConversion.h

Added a View menu for hiding/showing the regular sidebar and the new download info sidebar.

### CODE REVIEW

File: CefSharp.Example/CefExample.cs [hunk 1]
```diff
@@ -24,6 +24,7 @@ public static class CefExample
         public const string BasicSchemeTestUrl = "custom://cefsharp/SchemeTest.html";
         public const string ResponseFilterTestUrl = "custom://cefsharp/ResponseFilterTest.html";
         public const string DraggableRegionTestUrl = "custom://cefsharp/DraggableRegionTest.html";
+        public const string DownloadHandlerTestUrl = "custom://cefsharp/DownloadHandlerTest.html";
```
Comment (amaitland) [inline CefSharp.Example/CefExample.cs]: Where is this used?

File: CefSharp.Wpf.Example/Views/BrowserTabView.xaml [hunk 2]
```diff
@@ -14,7 +14,7 @@
         <BooleanToVisibilityConverter x:Key="BooleanToVisibilityConverter" />
         <ObjectDataProvider x:Key="BitmapScalingModeEnum" MethodName="GetValues" ObjectType="{x:Type system:Enum}">
             <ObjectDataProvider.MethodParameters>
-                <x:Type TypeName="BitmapScalingMode"/>
+                <x:Type TypeName="BitmapScalingMode" />
```
Comment (amaitland) [inline CefSharp.Wpf.Example/Views/BrowserTabView.xaml]: Please revert whitespace changes

File: CefSharp.Wpf.Example/Views/BrowserTabView.xaml [hunk 3]
```diff
@@ -44,8 +44,7 @@
                          GotKeyboardFocus="OnTextBoxGotKeyboardFocus"
                          GotMouseCapture="OnTextBoxGotMouseCapture">
                     <TextBox.InputBindings>
-                        <KeyBinding Key="Enter"
-                                    Command="{Binding GoCommand}" />
+                        <KeyBinding Key="Enter" Command="{Binding GoCommand}" />
```
Comment (amaitland) [inline CefSharp.Wpf.Example/Views/BrowserTabView.xaml]: Please revert whitespace changes

File: CefSharp.Wpf.Example/Views/BrowserTabView.xaml.cs [hunk 4]
```diff
@@ -85,6 +89,26 @@ public BrowserTabView()
...
+        }
+
+        private void OnDownloadUpdatedFired(object sender, DownloadItem e)
+        {
+            this.UpdateDownloadAction("OnDownloadUpdated", e);
+        }
+
+        private void UpdateDownloadAction(string downloadAction, DownloadItem downloadItem)
+        {
+            this.Dispatcher.Invoke(() =>
```
Comment (amaitland) [inline CefSharp.Wpf.Example/Views/BrowserTabView.xaml.cs]: Is there a reason to update in a `sync` fashion? `UI` updates should be `async` where possible.

File: .gitignore [hunk 5]
```diff
@@ -16,6 +16,8 @@
 *.swp
 *.user
 *.xml
+*.VC.opendb
```
Comment (amaitland) [inline .gitignore]: Would ideally be in it's own `PR`, though leave it for now. (Doesn't relate to the feature your adding).

File: CefSharp.Example/DownloadHandler.cs [hunk 6]
```diff
@@ -2,12 +2,23 @@
...
 {
     public class DownloadHandler : IDownloadHandler
     {
+        public event EventHandler<DownloadItem> OnBeforeDownloadFired;
+
+        public event EventHandler<DownloadItem> OnDownloadUpdatedFired;
+
         public void OnBeforeDownload(IBrowser browser, DownloadItem downloadItem, IBeforeDownloadCallback callback)
         {
+            if (this.OnBeforeDownloadFired != null)
```
Comment (amaitland) [inline CefSharp.Example/DownloadHandler.cs]: Would prefer firing of event handlers to look like below for consistency and best practice.

```c#
var handler = LoadingStateChanged;
if (handler != null)
{
    handler(this, args);
}
```

I know this is a demo, though I prefer to take a reference to a handler before executing as it's possible in a multithreaded environment for the handler reference to be cleared between the null check and executing. There's no

File: CefSharp.Example/DownloadHandler.cs [hunk 7]
```diff
@@ -19,7 +30,10 @@ public void OnBeforeDownload(IBrowser browser, DownloadItem downloadItem, IBefor
 
         public void OnDownloadUpdated(IBrowser browser, DownloadItem downloadItem, IDownloadItemCallback callback)
         {
-            
+            if (this.OnDownloadUpdatedFired != null)
```
Comment (amaitland) [inline CefSharp.Example/DownloadHandler.cs]: For consistency `this` has typically only been used when referencing a private variable.

File: CefSharp.Example/CefExample.cs (hunk 1 above)
Comment (philipbeber) [inline CefSharp.Example/CefExample.cs]: Sorry, meant to revert this. I had added a new test html file but it didn't turn out to be a very good addition.

File: CefSharp.Wpf.Example/Views/BrowserTabView.xaml (hunk 2 above)
Comment (philipbeber) [inline CefSharp.Wpf.Example/Views/BrowserTabView.xaml]: Done!

File: CefSharp.Wpf.Example/Views/BrowserTabView.xaml (hunk 3 above)
Comment (philipbeber) [inline CefSharp.Wpf.Example/Views/BrowserTabView.xaml]: Done!

File: .gitignore (hunk 5 above)
Comment (philipbeber) [inline .gitignore]: Okay!

File: CefSharp.Example/DownloadHandler.cs (hunk 6 above)
Comment (philipbeber) [inline CefSharp.Example/DownloadHandler.cs]: I update to match that style. Presumably using a null-conditional would also be okay? E.g.

`OnBeforeDownloadFired?.Invoke(this, downloadItem);`

**Comment by philipbeber** [comment]
Thanks for the feedback Alex! I have pushed fixes for those problems.

**Comment by amaitland** [comment]
Thanks, merged 👍

**Comment by philipbeber** [comment]
Awesome, thanks Alex!

**Pull request merged**